File: python/day2.py
from solution import Solution


class Day2(Solution):
    @classmethod
    def _Part1(cls) -> int:
        total = 0
        maxColors = {"red": 12, "green": 13, "blue": 14}

        for i, game in enumerate(cls.inputLines):
            if any(
                int(count) > maxColors[color]
                for count, color in [
                    s.strip().split(" ")
                    for r in game[game.find(":") + 2 :].split(";")
                    for s in r.split(",")
                ]
            ):
                continue

            total += i + 1

        return total

    @classmethod
    def _Part2(cls) -> int:
        total = 0
        for game in cls.inputLines:
            maxUsed = {}
            for r in game[game.find(":") + 2 :].split(";"):
                for s in r.split(","):
                    count, color = s.strip().split(" ")
                    maxUsed[color] = max(maxUsed[color], int(count))

            total += maxUsed["red"] * maxUsed["green"] * maxUsed["blue"]

        return total


Day2.Run("day2.txt")

File: rust/src/days/day2.rs
use crate::days::solution::Solution;
use std::{cmp::max, collections::HashMap};

pub struct Day2;

impl Solution for Day2 {
    fn part1(&self, input: &str) -> String {
        let mut total = 0;
        let max_colors = HashMap::from([("red", 12), ("green", 13), ("blue", 14)]);

        for (i, game) in self.input_lines(input).iter().enumerate() {
            if game
                .split(": ")
                .nth(1)
                .unwrap()
                .split(";")
                .flat_map(|r| r.split(','))
                .map(|s| s.trim().split(' ').collect::<Vec<_>>())
                .any(|count| count[0].parse::<i32>().unwrap() > max_colors[count[1]])
            {
                continue;
            }

            total += i + 1;
        }

        return total.to_string();
    }

    fn part2(&self, input: &str) -> String {
        let mut total = 0;
        for game in self.input_lines(input) {
            let mut max_used = HashMap::new();
            game.split(": ").nth(1).unwrap().split(";").for_each(|r| {
                r.split(',')
                    .map(|s| s.trim().split(' ').collect::<Vec<_>>())
                    .for_each(|count| {
                        let color = count[1];
                        let num = count[0].parse::<i32>().unwrap();
                        let current = max_used.entry(color).or_insert(0);
                        *max_used.entry(color).or_insert(0) = max(*current, num);
                    });
            });
            total += max_used.values().product::<i32>();
        }

        return total.to_string();
    }
}
